Facenet based on inception v1

AnnGareeva opened the pull request:
*  prototype of facenet
--> unresolved issues with weights files
--> as a result - can't check the quality of algorithm (now it
doesn't work correctly)
--> need to add the assertions and exceptions processing (check weights,
image sizes and other)
--> to do ideas from TODO comments
--> add a file env info

AnnGareeva committed:
Facenet based on inception v1

*  prototype of facenet
--> unresolved issues with weights files
--> as a result - can't check the quality of algorithm (now it
doesn't work correctly)
--> need to add the assertions and exceptions processing (check weights,
image sizes and other)
--> to do ideas from TODO comments
--> add a file env info

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,382 @@
+
+        input_size = Input(shape=(96, 96, 3))
+
+        self.conv1 = Conv2D(filters=64, kernel_size=(7, 7), strides=(2, 2),
```
You have not to store reference to all layers inside yours class because it will break garbage  collection.

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,382 @@
+                Activation('relu', name=f'inception_{addition_name}_1x1_act')(
+                    getattr(self, f'inception_{addition_name}_1x1_bn')))
+
+    def create_interception_block(self, input_layer, addition_name='',
```
`create_inception_block`, but I propose to split this function into multiple functions, like _inception_block_1a_, _inception_block_1b_. etc, because it significantly improves readability.

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,382 @@
+                                    getattr(self, f'inception_{addition_name}_pool_act')])
+
+        if conv1_5x5 is not None and conv2_5x5 is not None:
+            setattr(self, f'inception_{addition_name}_5x5_conv1',
+                          Conv2D(filters=conv1_5x5,
+                                 kernel_size=(1, 1),
+                                 padding='same',
+                                 kernel_regularizer=l2(0.01),
+                                 name=f'inception_{addition_name}_5x5_conv1')(input_layer))
+            setattr(self, f'inception_{addition_name}_5x5_bn1',
+                    BatchNormalization(name=f'inception_{addition_name}_5x5_bn1')(
+                        getattr(self, f'inception_{addition_name}_5x5_conv1')))
+            setattr(self, f'inception_{addition_name}_5x5_act1',
+                    Activation('relu', name=f'inception_{addition_name}_5x5_act1')(
+                        getattr(self, f'inception_{addition_name}_5x5_bn1')))
```
_setattr_ and _getattr_ should be used only for introspection tasks. You don't need it here.

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,382 @@
+                layer = self.model.get_layer(name=layer_name)
+                src_weights = layer.get_weights()
+                weights_list = [model_weights_by_layers[filename] for filename in files_names]
+                weights_list[0] = weights_list[0].reshape(src_weights[0].shape)
+                layer.set_weights(weights_list)
```
It will not work as expected, because weights are stored as NCHW tensor (or CoCiHW in terms of TensorFlow) with shape specified in the task Appendix. Therefore, you have to firstly reshape your tensor and physically transpose it to HWCiCo after that.

achernigin1987 replied:
I don't see transposing here. You just need to do the following:
```
conv_weights = np.reshape(conv_weights , conv_shape[name])
conv_weights  = np.transpose(conv_weights , (2, 3, 1, 0))
```

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,382 @@
+        Create plot with model graph
+        :return: None
+        """
+        plot_model(self.model, to_file='model.png', show_shapes=True, show_layer_names=True)
```
It will force me to setup _pydot_ package to my virtual environment. Could you please make it optional and disable by default ?

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,382 @@
+
+
+if __name__ =='__main__':
+    os.environ["PATH"] += os.pathsep + 'C:/Program Files (x86)/Graphviz2.38/bin/'
```
remove it

AnnGareeva replied:
accept

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,382 @@
+    print(res)
+
+    #who is the particular person and comparing linear distance between encodings with threshold=0.7.
+    res = encoder.model_evaluation(data, threshold=0.7)
```
It seems you use the same data to for database and validation

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,382 @@
+        # Common last layers
+        avg = GlobalAveragePooling2D(name='avg_pool')(self.inception_5b)
+        self.flatten_layer = Flatten()(avg)
+        self.dense_layer = Dense(units=128, kernel_regularizer=l2(0.01))(self.flatten_layer)
```
You don't need regularization for inference

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,382 @@
+                                                           pool_type='max')
+
+        # Common last layers
+        avg = GlobalAveragePooling2D(name='avg_pool')(self.inception_5b)
```
_GlobalAvgPooling2D_ returns single number per channel

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,382 @@
+                       kernel_size=(1, 1),
+                       padding='same',
+
+                       kernel_regularizer=l2(0.01),
```
remove regularizers everywhere

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,382 @@
+        :param addition_pool: add a stream with MaxPooling layer
+        :return: None
+        """
+        layers_set_list = []
```
should be removed

AnnGareeva replied:
Actually, I don't understand the reason. It's used for collections of output layers for following concatenating.

achernigin1987 replied:
If you split this very large function, then you will be be able to do smth like this:
```
x1 = Conv1(X)
x2 = BN(X1)
x3_1 = Activation(x2)
...
return concatenate([x3_1, x3_2, ..])
```
So, the main reason here is to improve readability. We have well-defined operations number in the each inception block, so it's better to not use generic approach inside blocks .

AnnGareeva committed:
facenet inception prototype

* fixed order (chw) in layers and in image array
! warning! it works only on GPU
* deleted function with model scheme creation
* refactored functions of inception blocks creation
* deleted str with command(adding to path)
* deleted normalization
* fixed the last pooling layer

--> model creation -> success
--> weights load -> success
--> making prediction -> success
! it works if you fix your weights files !

achernigin1987 commented:
Just remove files inception_5a_3x3_conv2_m.csv, inception_5a_3x3_conv2_v.csv from the weights directory. I don;t know why they exist there

achernigin1987 commented on model_v4.py:
```diff
@@ -0,0 +1,356 @@
+    encoder = FaceEncoder()
+    encoder.load_weights(args.weights)
+
+    indata_dict = prepare_data(args.input_data, img_size=(96, 96, 3))
+    res = encoder.data_encoding(indata_dict)
+    print(res)
+
+    res = encoder.model_evaluation(indata_dict, threshold=0.7)
```
still use the same data for database creation and validation
